### ISSUE TITLE
Make the crate self-contained by default?

### ISSUE THREAD

**Issue by perfah** [perfah/wlroots-rs#2]
Investigate whether it would be possible to make the crate fully self-contained by:

- defaulting to static linking

- making dependency libraries optional (such as pixman), and only explicitly enabled

- adding compilation messages to communicate what libraries must be installed if some dependency libraries are referenced dynamically

Is it worth it?

**Comment by gynther-k** [comment]
With my latest commit to issue/2 i tried to make pixman explicity enabled with
**cargo build --features withpixman**
If not provided it willbuild without. Do you think this is the way to go? Or should we do something different?

**Comment by gynther-k** [comment]
And I also had to fork issue/1 into this to make it build correctly. @perfah

**Comment by perfah** [comment]
@gynther-k Nice work! I only have two suggestions:

- change "withpixman" to simply "pixman" as this is the way it is usually done
- Avoid code duplication by doing the if-statement at the end instead:
    ```
    if cfg!(feature = "withpixman") {
        builder = builder.clang_arg("-DWLR_USE_PIXMAN");
    }
    ```
Are you saying that the project compiles with the "withpixman" flag only if pixman is installed and without regardless, just as intended? Merging issue/1 into this one is no problem!

**Comment by gynther-k** [comment]
@perfah Thanks. I Removed code duplication and changed to pixman instead. Yes it only compiles with pixman if you have it installed otherwise it fails. But it will then succeed without the pixman flag.

**Comment by gynther-k** [comment]
@perfah with the latest commit I can compile with `cargo build --features static` with the procedure below. But i had to add the wayland submodule in order for it to work with wlroots 0.10.0. Also some old static links needed to be removed. 

I added the wayland submodule with meson but there are still many dependencies that do not support meson.

**Comment by gynther-k** [comment]
```
git clone https://github.com/perfah/wlroots-rs.git
git checkout issue/2
git submodule update --init

sudo apt-get install 
cmake 
doxygen 
libffi-dev 
xmlto 
wayland-protocols 
texlive-font-utils 
xsltproc 
graphviz 
libxml2-dev
python3-pip
cargo
libwayland-bin 
llvm 
clang 

pip3 install --user ninja
pip3 install --user meson
export PATH="/home/{$USERNAME}/.local/bin:$PATH" 

cd wlroots-sys
cargo build --features static
```
should we add some of this to the Readme?

**Comment by perfah** [comment]
> I added the wayland submodule with meson but there are still many dependencies that do not support meson.

Okay, I am assuming you mean in dynamic mode. Either we build these libraries with something else, or just leave an error message in the terminal saying that this is a required dependency (see below). What do you think? If we do it in the former way, it could look something like this:

```
if !Path::new("xwayland").exists(){
    git.clone(..);
    build();
}
```

```
#ifdef STATIC
#include "xwayland/xwayland.h"
#else
#include <xwayland.h>
#endif 

```

>should we add some of this to the Readme?

I think we can add anything that is not exclusive to Ubuntu to the readme, maybe something like this:

```
git clone https://github.com/perfah/wlroots-rs.git
git checkout master
git submodule update --init

Required dependencies:
    meson
    wayland
    wayland-protocols
    EGL
    GLESv2
    libdrm
    GBM
    libinput
    xkbcommon
    udev
Optional dependencies:
    pixman
    systemd
    elogind
    libcap
    ...
```

**Comment by gynther-k** [comment]
I mean on the `cargo build --features static` where i needed to add the following.
```
    if Path::new("wayland").exists() {
        meson::build("wayland", build_path_str);
    } else {
        panic!("The `wayland` submodule does not exist");
    }
```
Otherwise I get the:
`meson.build:98:0: ERROR: Invalid version of dependency, need 'wayland-server' ['>=1.18'] found '1.16.0'.` because I think that the new wlroots 0.10.0 needs the newer version.

Okay then I understand, need to think a little about the other dependencies then....

**Comment by perfah** [comment]
I think maybe wayland should be one of the required dependencies that we need both in static and dynamic mode. Maybe we should just create a nice error messages if any of the following packages are not found (to avoid excessive work):

```
Required dependencies:
    meson
    wayland
    wayland-protocols
    EGL
    GLESv2
    libdrm
    GBM
    libinput
    xkbcommon
    udev 
```

**Comment by gynther-k** [comment]
@perfah with my latest commit 42fbc4b i have added a simple dependency check for the **default** `cargo build`. Which checks if the correct version of libraries is located in PATH or pkg-config. (I have created the print outs for inexperienced users so maybe they are a little overexplained?)

Do you think I should continue with the following?:
```
unstable
static
code cleaning 
```

**Comment by perfah** [comment]
@gynther-k I do like the idea of explaining in the way you describe but think we can have a call to clear things up. Here are some thoughts I have at the moment:

- do we need to check for version `1.18`? Isn't  the rust package `wayland-scanner = "0.25.*"` already guaranteed from the `cargo.toml` of `wlroots_sys`?

```
    if check_version(
        "wayland-scanner".to_string(),
        "--version".to_string(),
        "1.18".to_string()
    )
```

- I don't think we should **hardcode** the version of dynamically linked libraries either (along with Rust packages besides from what is defined the `cargo.toml`). Only print good error messages when a problem occurs (if possible). 

- The most important thing is that when introducing this crate you shouldn't get tons of weird errors in the console. One way of achieving this is by defaulting (through cargo features) to a state where most dependencies are taken cared of and when they do fail it is explained clearly (I think you are on the right track there)

I can't really answer on the question of whether I think you should continue. It depends on how you are thinking.

**Comment by gynther-k** [comment]
When cloning issue/2 (with dependecy check removed) and running the following on a clean installation of ubuntu get the following error:
```
sudo apt-get install cargo
sudo apt-get install pkg-config

cargo build
```
 ```
  Compiling wlroots-sys v0.4.0 (/home/gynther/Desktop/winshare/apr22/wlroots-rs/wlroots-sys)
error: failed to run custom build command for `wlroots-sys v0.4.0 (/home/gynther/Desktop/winshare/apr22/wlroots-rs/wlroots-sys)`

Caused by:
  process didn't exit successfully: `/home/gynther/Desktop/winshare/apr22/wlroots-rs/target/debug/build/wlroots-sys-585bcc1cba01567c/build-script-build` (exit code: 101)
--- stderr
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Failure { command: "\"pkg-config\" \"--variable=prefix\" \"wayland-protocols\"", output: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "Package wayland-protocols was not found in the pkg-config search path.\nPerhaps you should add the directory containing `wayland-protocols.pc\'\nto the PKG_CONFIG_PATH environment variable\nNo package \'wayland-protocols\' found\n" } }', src/libcore/result.rs:1188:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

```
It seems like the error occures here:

```
meson();
let protocol_header_path =
generate_protocol_headers().expect("Could not generate header files for wayland protocols");

let protocols_prefix = pkg_config::get_variable("wayland-protocols", "prefix").unwrap();


```
This is solved by running:

`sudo apt-get install wayland-protocols.
`

or manual install

```
git clone wayland-scanner
install...
git clone wayland-protocols
install...
```
Because I wanted to add an instruction for the inexperienced user
to install manually (if they have a system with bad/no packet manager). 

Because manual install of wayland-protocols need dependency wayland-scanner, 
then it doesn't help that I have the wayland-scanner = "0.25.*" in Cargo.toml.
And I took the latest version 1.18 at github.

* Do you get the same result when running on a clean installation of ubuntu?

* Should we have instructons for installing wayland-protocols manually(if we have bad packet manager) ?

* Should I run the dependency check only with a `cargo build features --helpful` flag instead? 

* Or can you do a small commit to guide me in the right direction?

**Comment by gynther-k** [comment]
fixed with 70351ac
